File: game/generation/astar.py
"""Calculates the shortest path from one point to another using the A* algorithm."""
from __future__ import annotations

# Builtin
from heapq import heappop, heappush

# Pip
import numpy as np

# Custom
from game.common import grid_bfs
from game.constants.generation import TileType
from game.generation.primitives import Point

__all__ = ("calculate_astar_path",)


def heuristic(a: Point, b: Point) -> int:
    """Calculate the Manhattan distance between two points.

     This preferable to the Euclidean distance since we can generate staircase-like
     paths instead of straight line paths.

     Further reading which may be useful:
     `Manhattan distance <https://en.wikipedia.org/wiki/Taxicab_geometry>`_
     `Euclidean distance <https://en.wikipedia.org/wiki/Euclidean_distance>`_

    Parameters
    ----------
    a: Point
        The first point.
    b: Point
        The second point.
    """
    return abs(a.x - b.x) + abs(a.y - b.y)


def calculate_astar_path(grid: np.ndarray, start: Point, end: Point) -> list[Point]:
    """Calculate the shortest path from one point to another using the A* algorithm.

    Further reading which may be useful:
    `The A* algorithm <https://en.wikipedia.org/wiki/A*_search_algorithm>`_

    Parameters
    ----------
    grid: np.ndarray
        The 2D grid which represents the dungeon.
    start: Point
        The start point for the algorithm.
    end: Point
        The end point for the algorithm.

    Returns
    -------
    list[Point]
        A list of points mapping out the shortest path from start to end.
    """
    # Initialise a few variables needed for the pathfinding
    heap: list[tuple[int, Point]] = [(0, start)]
    came_from: dict[Point, Point] = {start: start}
    distances: dict[Point, int] = {start: 0}
    total_costs: dict[Point, int] = {start: 0}

    # Loop until the heap is empty
    while heap:
        # Get the lowest-cost point from the heap
        _, current = heappop(heap)

        # Check if we've reached our target
        if current == end:
            # Backtrack through came_from to get the path and return the result
            result = []
            while True:
                # Add the path to the result list
                result.append(current)

                # Test if we've hit the starting point
                if came_from[current] != current:
                    current = came_from[current]
                else:
                    break
            return result

        # Add all the neighbours to the heap with their cost being f = g + h:
        #   f - The total cost of traversing the neighbour.
        #   g - The distance between the start point and the neighbour point.
        #   h - The estimated distance from the neighbour point to the end point.
        for bfs_neighbour in grid_bfs(current, *grid.shape):
            if bfs_neighbour not in came_from:
                # Store the neighbour's parent and calculate its distance from the start
                # point
                neighbour = Point(*bfs_neighbour)
                came_from[neighbour] = current
                distances[neighbour] = distances[came_from[neighbour]] + 1

                # Check if the neighbour is an obstacle
                if grid[neighbour.y][neighbour.x] == TileType.OBSTACLE:
                    # Set the total cost for the obstacle to infinity
                    total_costs[neighbour] = np.inf
                else:
                    # Set the total cost for the neighbour to f = g + h
                    total_costs[neighbour] = distances[neighbour] + heuristic(
                        current, neighbour
                    )

                # Add the neighbour to the heap
                heappush(heap, (total_costs[neighbour], neighbour))

    # A path can't be found
    return []

File: src/hades/extensions/src/astar.cpp
// Definitions
#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION

// Includes
#include "hades_common.h"
#include <numpy/arrayobject.h>
#include <queue>
#include <iostream>


// ----- CONSTANTS ------------------------------
/* The ID of a TileType obstacle */
int OBSTACLE_ID;


// ----- DOCSTRINGS ------------------------------
/* Stores the docstring for the astar module */
PyDoc_STRVAR(
    astar_module_docstring,
    "Calculate the shortest path in a grid from one pair to another using the A* "
    "algorithm.\n\n"
);


/* Stores the docstring for the calculate_astar_path method */
PyDoc_STRVAR(
    astar_docstring,
    "Calculate the shortest path in a grid from one pair to another using the A* "
    "algorithm.\n\n"
    "Further reading which may be useful:\n"
    "`The A* algorithm <https://en.wikipedia.org/wiki/A*_search_algorithm>`_\n\n"
    "Parameters\n"
    "----------\n"
    "grid: np.ndarray\n"
    "   The 2D grid which represents the dungeon.\n"
    "start: Pair\n"
    "    The start pair for the algorithm.\n"
    "end: Pair\n"
    "    The end pair for the algorithm.\n\n"
    "Returns\n"
    "-------\n"
    "list[Pair]\n"
    "    A list of pairs mapping out the shortest path from start to end."
);


// ----- STRUCTURES ------------------------------
struct Neighbour {
    /* Represents a namedtuple describing a grid pair and its cost from the start
    position */
    int cost;
    IntPair pair;

    inline bool operator<(const Neighbour nghbr) const {
        // The priority_queue data structure gets the maximum priority, so we need to
        // override that functionality to get the minimum priority
        return cost > nghbr.cost;
    }
};


// ----- PYCFUNCTIONS ------------------------------
static PyObject *calculate_astar_path(PyObject *self, PyObject *args) {
    /* Parse arguments */
    PyArrayObject *grid;
    struct IntPair start, end;
    if (!PyArg_ParseTuple(args, "O(ii)(ii)", &grid, &start.x, &start.y, &end.x, &end.y)) {
        Py_RETURN_NONE;
    }

    // Do some validation checking on the given parameters
    if (PyArray_Check(grid) == 0) {
        Py_RETURN_NONE;
    }

    /* A* algorithm logic */
    // Set up a few variables needed for the pathfinding
    PyObject *result = PyList_New(0);
    std::priority_queue<Neighbour> queue;
    queue.push({0, start});
    std::unordered_map<IntPair, IntPair> came_from = {{start, start}};
    std::unordered_map<IntPair, int> distances = {{start, 0}};
    int height = (int) PyArray_DIM(grid, 0);
    int width = (int) PyArray_DIM(grid, 1);

    // Loop until the priority queue is empty
    while (!queue.empty()) {
        // Get the lowest cost pair from the priority queue
        IntPair current = queue.top().pair;
        queue.pop();

        // Check if we've reached our target
        if (current == end) {
            // Backtrack through came_from to get the path
            while (!(came_from[current] == current)) {
                // Add the current pair to the result list
                PyList_Append(result, Py_BuildValue("(ii)", current.x, current.y));

                // Get the next pair in the path
                current = came_from[current];
            }

            // Add the start pair and exit out of the loop
            PyList_Append(result, Py_BuildValue("(ii)", start.x, start.y));
            break;
        }

        // Add all the neighbours to the heap with their cost being f = g + h:
        //   f - The total cost of traversing the neighbour.
        //   g - The distance between the start pair and the neighbour pair.
        //   h - The estimated distance from the neighbour pair to the end pair. We're using the Manhattan distance for
        //       this.
        for (IntPair neighbour: grid_bfs(current, height, width)) {
            if (!came_from.count(neighbour)) {
                // Store the neighbour's parent and calculate its distance from the
                // start pair
                came_from[neighbour] = current;
                distances[neighbour] = distances[current] + 1;

                // Check if the neighbour is an obstacle
                int f_cost;
                if (*((int *) PyArray_GETPTR2(grid, neighbour.y, neighbour.x)) == OBSTACLE_ID) {
                    // Set the total cost for the obstacle to infinity
                    f_cost = INT_INFINITY;
                } else {
                    // Set the total cost for the neighbour to f = g + h
                    f_cost = distances[neighbour] + (abs(current.x - neighbour.x) + abs(current.y - neighbour.y));
                }

                // Add the neighbour to the priority queue
                queue.push({f_cost, neighbour});
            }
        }
    }

    /* Return result */
    return result;
}


// ----- METHOD DEFINITIONS ------------------------------
static PyMethodDef astar_methods[] = {
    /* Defines the metadata for methods accessible through Python */
    {"calculate_astar_path", calculate_astar_path, METH_VARARGS, astar_docstring},
    {NULL},
};


// ----- MODULE DEFINITION ------------------------------
static struct PyModuleDef astar_module = {
    /* Defines the metadata for this extension module */
    PyModuleDef_HEAD_INIT,
    "astar",
    astar_module_docstring,
    -1,
    astar_methods,
};


// ----- MODULE CREATION ------------------------------
PyMODINIT_FUNC PyInit_astar(void) {
    /* Initialises this module so Python can access it */
    // Initialise the C++ module and check if it's valid or not
    PyObject *module = PyModule_Create(&astar_module);
    if (module == NULL)
        return NULL;

    // Initialise the constants
    PyObject *temp_obstacle_id = get_global_constant("hades.constants.generation", {"TileType", "OBSTACLE", "value"});
    if (temp_obstacle_id == Py_None) {
        return NULL;
    } else {
        OBSTACLE_ID = (int) PyLong_AsLong(temp_obstacle_id);
    }
    Py_DECREF(temp_obstacle_id);

    // Initialise the Numpy C-API
    import_array();

    // Return the C++ initialised module
    return module;
}
